File: 41-first_missing_positive.py
def firstMissingPositive(nums):
    """
    This function finds the first missing positive integer in an unsorted list.
    It uses an in-place cyclic sort to achieve O(n) time complexity.
    """
    
    n = len(nums)
    
    # Step 1: Place each number in its correct position (nums[i] -> index nums[i] - 1)
    for i in range(n):
        while 1 <= nums[i] <= n and nums[i] != nums[nums[i] - 1]: 
            # Swap nums[i] with its correct position nums[nums[i] - 1]
            nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]

    # Step 2: Find the first missing positive number
    for i in range(n):
        if nums[i] != i + 1:
            return i + 1  # First missing positive integer
    
    # Step 3: If all numbers are in place, return n + 1
    return n + 1

File: 41-first_missing_positive.js
var firstMissingPositive = function (nums) {
    let n = nums.length;

    // Step 1: Place each number in its correct position (i.e., nums[i] should be i + 1)
    for (let i = 0; i < n; i++) {
        while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] !== nums[i]) {
            // Swap nums[i] with nums[nums[i] - 1]
            let temp = nums[i];
            nums[i] = nums[temp - 1];
            nums[temp - 1] = temp;
        }
    }

    // Step 2: Identify the first missing positive number
    for (let i = 0; i < n; i++) {
        if (nums[i] !== i + 1) {
            return i + 1;
        }
    }

    // If all numbers are in their correct positions, return n + 1
    return n + 1;
};

// Example usage:
console.log(firstMissingPositive([3, 4, -1, 1])); // Output: 2
console.log(firstMissingPositive([1, 2, 0])); // Output: 3
console.log(firstMissingPositive([7, 8, 9, 11, 12])); // Output: 1
